### DISCUSSION TITLE
Support of `assert {type: "json"}`

### DISCUSSION THREAD

**Discussion by pubmikeb** [eslint/eslint#15305] [Q&A] [answered]
Starting Node.js 17.1.x, the `assert {type: "json"}` suffix is mandatory to be able importing the JSON modules. ESLint 8.2.0 is not aware of this syntax and therefore marks it as an error: 

> ESLint: Parsing error: Unexpected token assert

The problem is that after this error is thrown, no further linting is performed, ESLint doesn't analyze any further line of code and bearing in mind that usually loading of JSON modules defined somewhere at the beginning of a file, ESLint doesn't scan the entire file.

Currently, the [Import Assertions](https://github.com/tc39/proposal-import-assertions) is on the stage 3 and I know that ESLint accepts new JS features starting the stage 4, but bearing in mind a fact that without a support of `assert {type: "json"}`, ESLint can't scan the entire file, it would be great to propose some workaround prior reaching the stage 4.

**Comment by nzakas** [comment, marked as answer]
We have the policy to wait until stage four because We don’t know which ECMA version it will end up in. Once import assertions have landed in an ECMAScript version, we can look at adding support in ESLint. In the meantime, you’ll need to use some thing like the Babel ESLint parser to get the support.

**Reply by dsine-de** [reply]
Since the latest LTS and the current version of Node.js (16.14.0 & 17.x) both require the `assert` syntax for JSON imports, I'd like to know if there is another way to use ESLint without adding `babel` - is it possible with a plugin? (Probably not, because parsing happens before any plugins).

Comments like `// eslint-disable-line` or `// eslint-disable-next-line` also don't work for the parser so we can't use ESLint with Node.js currently (without babel, which most people probably don't want to add to Node.js code because it doesn't need to get transformed).

I undestand the policy though.

**Reply by pubmikeb** [reply]
It's quite strange why Node.js decided to force usage of `assert {type: "json"}` prior reaching the stage 4.
But bearing in mind the current situation, I would consider adding support of `assert {type: "json"}` to ESLint already at the stage 3 to provide a smooth linting experience for the mandatory feature of the GA Node.js versions.

**Reply by dsine-de** [reply]
I (and probably many others) would really appreciate a solution (perhaps a `ecmaFeatures` configuration option or, if possible, an additional plugin we need to add as a dependency for the time being), thank you!

**Reply by Trott** [reply]
> It's quite strange why Node.js decided to force usage of `assert {type: "json"}` prior reaching the stage 4.

At this time, [it is an Experimental API](https://nodejs.org/dist/latest-v17.x/docs/api/esm.html#import-assertions). Using it in the most recent Node.js release results in this warning:

```
ExperimentalWarning: Importing JSON modules is an experimental feature. This feature could change at any time
```

That (to me, at least) argues for not shipping it in ESLint yet. It does seem like "install and use `@babel/eslint-parser`" is the right answer for folks using features like this.

**Reply by nzakas** [reply]
Our policy is only to ship stage 4 features. We don’t make exceptions because supporting pre-stage 4 features requires extra maintenance for little added value. You can use an alternative parser that supports experimental features in the meantime.

**Reply by aguegu** [reply]
But this feature is in LTS nodejs 16.14.0 now. https://nodejs.org/docs/latest-v16.x/api/errors.html#err_import_assertion_type_missing

**Reply by dsine-de** [reply]
@aguegu But the possibility to import JSON files as modules is still behind the `--experimental-json-modules` flag.

**Reply by aguegu** [reply]
True, but is there any walk around that can do about eslint that we still want it to enforce style.

**Reply by aguegu** [reply]
I have tried another approach that instead of the `static` import, maybe eslint can work with [`dynamic` import](https://v8.dev/features/import-assertions#dynamic-import()). With node 16.14.0, it has to be like this 

```
await import('../../package.json', { assert: { type: 'json' } });
```

But then I got an `Parsing error: Unexpected token ,`. My use case for json file import is simple, all I need is the version number in `package.json` that maintained by `npm version`. 

My setup is eslint@8.9.0 with
```
parserOptions: {
  ecmaVersion: 'latest',
  sourceType: 'module',
}
``` 
and without babel.

**Reply by thescientist13** [reply]
FWIW, Node 18 is now LTS and although JSON Modules is still experimental, it is not behind a flag anymore.
https://nodejs.org/docs/latest-v18.x/api/esm.html#json-modules

**Reply by gaving** [reply]
Regularly refresh this issue to see if/when this will be supported now that we've moved to LTS, not particularly keen on the workarounds..

**Reply by jammi** [reply]
Are you done waiting? I'm encountering this linting code running on node v19.3.0

**Reply by RagnarLothbrok-Odin** [reply]
^ what jammi said, this warning is irritating

**Reply by ntatko** [reply]
Still an issue...

**Reply by qubyte** [reply]
Worth noting: that while JSON modules are now their own proposal at stage 3, the proposal which spawned it (import assertions) had issues, dropped back to stage 2, and then was changed to [import attributes](https://github.com/tc39/proposal-import-attributes). That proposal replaces the keyword `assert` with `with`. I don't have an insight into the process, but the odds seem good that it will affect the JSON modules proposal.

It seems wise to be especially strict with this proposal.

**Reply by alexfernandez** [reply]
It's now 2023-08-16 and the [import proposal](https://github.com/tc39/proposal-import-attributes) seems to be back at stage 3. To be honest both options, `with` and `assert`, are horrible; but at this stage anything is better than nothing since it's already implemented in Node.js. Let's hope it's approved soon :crossed_fingers:

**Reply by sylver** [reply]
Oh, well.

![](https://i.imgflip.com/12fosi.jpg)

**Reply by Rajkumar7633** [reply]
import apiDocs from './swagger.json' assert { type: 'json' };
                                     ^^^^^^

SyntaxError: Unexpected identifier 'assert'
    at compileSourceTextModule (node:internal/modules/esm/utils:337:16)
    at ModuleLoader.moduleStrategy (node:internal/modules/esm/translators:163:18)
    at callTranslator (node:internal/modules/esm/loader:430:14)
    at ModuleLoader.moduleProvider (node:internal/modules/esm/loader:436:30)
    at async ModuleJob._link (node:internal/modules/esm/module_job:106:19)

Node.js v22.2.0
[nodemon] app crashed - waiting for file changes before starting... error are coming

**Reply by SilasCrosby** [reply]
Yo, why isn't assert working yet?

**Reply by pubmikeb** [reply]
> why isn't assert working yet?

It's still in the stage 3. Let's wait until the stage 4 will be reached.

BTW, `assert` has been [replaced](https://github.com/tc39/proposal-import-attributes?tab=readme-ov-file#synopsis) with `with`:

```js
import json from "./foo.json" with { type: "json" };
import("foo.json", { with: { type: "json" } });
```

**Reply by xavierfoucrier** [reply]
I can confirm **it's now well supported** since `eslint` 9.19.0 🎉

**Comment by msurdi** [comment]
I know it is far from ideal and only works on a node environment, but this workaround worked for me in the meanwhile and I thought it might help others reading this in the future:

```js
import { readFileSync } from "fs";
const packageJSON = JSON.parse(readFileSync("./package.json"));
```

**Reply by kenbanks-peng** [reply]
Thank you @msurdi. It seems to me that relative paths in import statements are relative to the file in which they are written, but relative paths in the parameter to readFileSync is relative to where node is run. I suppose in the case of a server, node is likely run from the same path location so it doesn't matter, but using node spuriously via a command line for running scripts from various path locations, this is an issue. Maybe better to ensure relative paths are relative to the file like import.

Seems __dirname is not supported in Node ES Modules, so URL and import.meta.url does the job for me. Maybe this could be improved further with import mapping or parameterizing it into a utility function importJson.

```
import { readFileSync } from "fs";
const fileUrl = new URL("../../path/relative/to/this/file/myJson.json", import.meta.url);
const packageJSON = JSON.parse(readFileSync(fileUrl));
```

**Reply by qubyte** [reply]
This blocks, when ES modules need not since they're async. Top-level await is available in all supported versions of Node, so the following is better:

```javascript
import { readFile } from 'node:fs/promises';
const fileUrl = new URL("<the relative path from this module>", import.meta.url);
const parsedPackageJSON = JSON.parse(await readFile(fileUrl, 'utf8'));
```

**Reply by leke-osez** [reply]
> This blocks, when ES modules need not since they're async. Top-level await is available in all supported versions of Node, so the following is better:
> 
> ```js
> import { readFile } from 'node:fs/promises';
> const fileUrl = new URL("<the relative path from this module>", import.meta.url);
> const parsedPackageJSON = JSON.parse(await readFile(fileUrl, 'utf8'));
> ```

**Comment by TriMoon** [comment]
@pubmikeb 
> https://github.com/eslint/eslint/discussions/15305#discussioncomment-2400923
>I know it is far from ideal and only works on a node environment, but this workaround worked for me in the meanwhile and I thought it might help others reading this in the future:

 **BIG thanks** to @msurdi  :100: :heart_eyes:  !!!
<details>
<summary>I now use this as my <kbd>rollup.config.mjs</kbd> thanks to you:</summary>

```js
import { copy } from '@guanghechen/rollup-plugin-copy'
import sourcemaps from 'rollup-plugin-sourcemaps'
import { terser } from 'rollup-plugin-terser'
/**
 * Work-around for eslint's failure for import-assert !
 * @todo Until eslint supports this, we need this work-around !
 * @see {@link https://github.com/eslint/eslint/discussions/15305#discussioncomment-2400923 solutions attribution}
 */
// import pkg from './package.json' assert {type: 'json'};
import { readFileSync } from 'fs'

function addCopyrights(minified){
	return [
		'/**'
		,(minified ? ' * minified version\n' : '')
			+ ' * @copyright	See the `/LICENSE.*` file for details.'
		,' * @license		CC-BY-NC-SA-4.0'
		,' */'
	].join('\n')
}

const
	buildDir = 'public'
	,sourcesDir = 'src'
	,_buildStamp = new Date().toISOString()
	/* eslint-disable comma-style */
	,/**
	 * Work-around for eslint's failure for import-assert !
	 * @todo Until eslint supports this, we need this work-around !
	 * @see {@link https://github.com/eslint/eslint/discussions/15305#discussioncomment-2400923 solutions attribution}
	 */
	pkg = JSON.parse(readFileSync('./package.json'))
	/* eslint-enable comma-style */

	,commonOutputOptions = {
		dir: buildDir
		,entryFileNames: '[name].mjs'
		,banner:
			addCopyrights(false)
			// Add _BUILDSTAMP_ here so it will be optimized also.
			+ `\nconst _BUILDSTAMP_ = "${pkg.version}@${_buildStamp}";`
		,preserveModules: false
		,preserveModulesRoot: sourcesDir
		,sourcemap: true
		,sourcemapExcludeSources: true
	}

	,terserOptions = {
		ecma: '2015'
		,format: {
			preamble: addCopyrights(true)
			// Don't preserve ANY comments.
			,comments: false
		}
		/**
		* Required for our logger to show meaningfull stacktraces !
		*/
		// eslint-disable-next-line camelcase
		,keep_fnames: true	// spell-checker: disable-line
		// eslint-disable-next-line camelcase
		,keep_classnames: true
		,module: true
		,compress: false
		// ,compress: {
		// 	passes: 2	// More as 2 has no effect atm...
		// }
	}

	,copyStaticFilesOptions = {
		verbose: true
		,targets: [
			// Unaltered
			{
				src: [
					'LICENSE.md'
					,'README.html'
					,`${sourcesDir}/*.{css,svg,png}`
				]
				,dest: buildDir
				// ,copyOnce: true
			}
			// Using non-minimized js
			,{
				src: `${sourcesDir}/*.html`
				,dest: buildDir
				,rename:	name => name + '.nonmin.html'
			}
			// Using minimized js
			,{
				src: `${sourcesDir}/*.html`
				,dest: buildDir
				// ,copyOnce: true
				,transform:	contents => contents
					.toString()
					.replace('app.mjs' ,'app.min.mjs')
			}
		]
	}

export default {
	input: [
		`${sourcesDir}/app.mjs`
		,`${sourcesDir}/ThreeD-app.mjs`
	]
	,output: [
		{
			...commonOutputOptions
		}
		,{
			...commonOutputOptions
			,entryFileNames: '[name].min.mjs'
			,plugins: [
				terser(terserOptions)
			]
		}
	]
	,plugins: [
		sourcemaps()
		,copy(copyStaticFilesOptions)
	]
	,watch: {
		clearScreen: false
	}
	// ,perf: true
}
```
</details>
Maybe that config can help some other readers in future also :wink:

**Comment by zhyd1997** [comment]
Try this solution: 
```bash
yarn add -D @babel/eslint-parser @babel/plugin-syntax-import-assertions
```
```js
// .eslintrc.js
  parser: "@babel/eslint-parser",
  parserOptions: {
    requireConfigFile: false,
    babelOptions: {
      plugins: [
        '@babel/plugin-syntax-import-assertions'
      ],
    },
  },
```
It works for me.

**Reply by johnsonsamuel** [reply]
Thank you so much for this 🙏🏽.

**Reply by gugiserman** [reply]
works like a charm! thanks! 🚒

**Reply by pprasannan** [reply]
Thank you so much @zhyd1997 ❤️

**Reply by qa-leads-videoslots** [reply]
I did the following and while it worked well, all of my rules are no longer being triggered, any help with that?

**Reply by thescientist13** [reply]
Yeah, my `plugins` don't seem to be working after this (though the import attributes syntax now lints correctly), but its only a minor one so I can live with it for now.

**Reply by Nafiunalamtasnim** [reply]
parserOptions: {
  sourceType: "module"
}

**Comment by neilpa** [comment]
If you control the launching of the node process (e.g. a test suite) you can avoid the import assertion completely with a [custom loader](https://nodejs.org/docs/latest-v18.x/api/esm.html#loaders). You do get some annoying warnings though.

```
$ node --experimental-loader ./loader.js main.js             
(node:55385) ExperimentalWarning: Custom ESM Loaders is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
{ foo: 'bar' }
(node:55385) ExperimentalWarning: Importing JSON modules is an experimental feature. This feature could change at any time
```

**main.js**
```js
import data from './data.json'
console.log(data);
```

**data.json**
```json
{
  "foo": "bar"
}
```

**loader.js**
```js
import path from 'path';
import { promises as fs } from 'fs';
import urlModule from 'url';

export async function load(url, context, defaultLoad) {
  const filename = urlModule.fileURLToPath(url);
  const parts = path.parse(filename);
  if (parts.ext === ".json") {
    const source = await fs.readFile(filename);
    return { source, format: 'json', shortCircuit: true };
  }

  return await defaultLoad(url, context);
}
```

**Reply by Nafiunalamtasnim** [reply]
node --no-warnings --experimental-loader ./loader.js main.js

**Comment by joostdecock** [comment]
I use this low-tech solution where I bundle all JSON imports in 1 file, and then re-export them as named exports.
Then, I just put this 1 file (that I call `json-loader.mjs` in `.eslintignore` and gone are the errors.

```mjs
/*
 * Eslint does not support import assertions because they only
 * support stage-4 language features.
 *
 * It's annoying and can't be disabled. So instead this file
 * will import all JSON and you can then import it from here.
 *
 * This way, we just ignore this file in eslint and voila.
 */
import dockerAllContainersExample from './docker-all-containers.json' with { type: 'json' }
import dockerRunningContainersExample from './docker-running-containers.json' with { type: 'json' }
import dockerDfExample from './docker-df.json' with { type: 'json' }
import dockerInfoExample from './docker-info.json' with { type: 'json' }
import dockerImagesExample from './docker-images.json' with { type: 'json' }
import dockerNetworksExample from './docker-networks.json' with { type: 'json' }


export {
  dockerAllContainersExample,
  dockerRunningContainersExample,
  dockerDfExample,
  dockerInfoExample,
  dockerImagesExample,
  dockerNetworksExample,
}
```

**Reply by mryechkin** [reply]
Clever! I'll have to remember this.

**Reply by phish108** [reply]
A little side note: `assert` has dropped from the spec. The correct syntax uses now `with`

```js
import foobarExample from './foobar.json' with { type: 'json' }
```

see: https://nodejs.org/api/esm.html#import-specifiers

**Reply by joostdecock** [reply]
Ah yes, good point. I've updated the example accordingly, for the benefit of future visitors.
Thanks for pointing it out 🙏

**Reply by iam-malc** [reply]
Future visitors would like to thank both @joostdecock and @phish108. Without your efforts, I would be lost ◡̈ 

EDIT: the change was moved to a different heading, see here: https://nodejs.org/api/esm.html#import-attributes

**Comment by thescientist13** [comment]
The day we've all been waiting for (😄 ), Import Attributes have now finally advanced to [Stage 4 at TC39](https://x.com/robpalmer2/status/1843491103120732585)! 🎉 
> Congrats to [@NicoloRibaudo](https://x.com/NicoloRibaudo) on advancing both Import Attributes & JSON Modules to Stage 4 at [@TC39](https://x.com/TC39) today 🎉
> 
> It introduces the `with` keyword for imports. Chrome successfully unshipped the `assert` keyword 👍

----

Not sure if we should introduce an issue for it now formally or if there is another method the team would desire for tracking this change?

**Reply by allstarschh** [reply]
I filed an issue for this, and turn out import attributes are supported already.
https://github.com/eslint/eslint/issues/19378

**Comment by xavierfoucrier** [comment]
### Now supported since ESLint 9.19.0 🎉

Hello 👋
I am still seeing a lot of comments about that and **just wanted to share that it's now well supported** since `eslint@9.19.0`.

```js
// import attributes syntax
import config from './src/config.json' with { type: 'json' };
```

Thanks to everyone that make it possible 🚀 😽

**Reply by Nafiunalamtasnim** [reply]
import { readFile } from 'fs/promises';
const config = JSON.parse(await readFile(new URL('./src/config.json', import.meta.url)));

**Comment by AAAWE1** [comment]
`import { defineConfig } from "eslint/config"
import { createRequire } from "module"

const require = createRequire(import.meta.url)
const autoImportConfig = require("./.eslintrc-auto-import.json")

export default defineConfig([
  {
    files: ["**/*.{js,mjs,cjs,vue}"],
    languageOptions: {
      ecmaVersion: 2022,
      sourceType: "module",
      globals: {
        ...autoImportConfig.globals,
      },
    },
  },
])
`
